docs/http-status-codes.adoc

```diff
@@ -0,0 +1,166 @@
+
+We opt for the first approach in most cases because then you can then fetch the Instance again and see its state, etc.
+
+In cases where the request does not create a new resource (e.g., "detach a disk"), we use "202 Accepted".
+
+=== Client errors vs. server errors
+
+When it comes to errors, **the most important distinction is between 400-level and 500-level status codes**.  https://httpwg.org/specs/rfc7231.html#status.codes[RFC 7231] summarizes the distinction:
+
+* 400-level codes (often called "client errors") mean "the request contains bad syntax or cannot be fulfilled".
+* 500-level codes (often called "server errors") mean "the server failed to fulfill an apparently valid request".
+
+Critically, a 500-level response means the problem is _completely outside the client's control_.  Common reasons include that an internal dependency is offline or overloaded or the server just hit a bug.  There's nothing the client can do except maybe retry, and even that isn't always appropriate.
+
+400-level codes don't necessarily represent a problem, mistake, or bug.  A client might attempt a conditional GET and get back a 400-level error saying the preconditions weren't true.  That might be totally expected on the client side under normal conditions.
+
+=== Client errors
+
+We use "400 Bad Request" for most types of invalid input.  This is pretty arbitrary.  Other popular choices include "409 Conflict" and "422 Unprocessible entity".  The spec for "409" really only seems to apply when the requested change conflicts with the underlying state, which is some kinds of invalid input (e.g., booting an Instance that's currently running) but not all.  422 comes from WebDAV and is not even mentioned in RFC 7231.
+
+=== Server errors
+
+Some clients erroneously treat all 500-level errors the same and retry them.  But it's useful to distinguish them:
+
+"500 Internal Server Error" basically means that the server hit a _bug_.  Retrying is not likely to be useful.  If an operator sees this, they should probably call support.  We might automatically open a support case when we see this.
+
+"503 Service Unavailable" means the server is _currently_ unable to handle the request -- usually this means something is overloaded or a dependency is not working.  Retrying _is_ likely a good idea.  On the operator side: the system should be providing the operator with information about known problems that might cause this, like sleds that are offline or disks that need to be replaced.  If the operator has resolved all the issues being reported them and they still see 503s, that's a bug!
+
+=== Other status codes
+
+There are also https://httpwg.org/specs/rfc7231.html#rfc.section.6.2[100-level status codes], whose use is generally more standardized and well-understood.
+
+There are also https://httpwg.org/specs/rfc7231.html#status.3xx[300-level status codes], which are pretty well-understood and we don't expect to make heavy use of.
```

Comment:
do we expect to make light use of them? Is there an area where we expect to use 3xx?

Reply:
I'm just hedging here again.  There's nothing wrong with them.  There are good reasons we might use them but I don't believe we use them today. (Example use: if you rename a resource, we could redirect requests to the old name to the new one until/unless the old name is replaced.  Or: we might want a resource to show up in more than one place in the hierarchy (e.g., a disk could show up under the /disks as well as the instance to which it's attached), and we might decide it's clearer to have one redirect to the other rather than pretend it's in both places.)

Maybe it would be clearer to just combine this with the previous paragraph to say "There are also 100-level and 300-level status codes, whose uses are more standardized and well-understood.  If we use these status codes, we should stick to the standard patterns."

Reply:
I did that in 01a7cb36.